core/node/node_framework/src/implementations/resources/pools.rs

```diff
@@ -1,175 +1,148 @@
         self.get_custom(1).await
     }
 
-    pub async fn get_custom(&self, size: u32) -> anyhow::Result<ConnectionPool<Core>> {
-        let result = self.builder.clone().set_max_size(size).build().await;
+    pub async fn get_custom(&self, size: u32) -> anyhow::Result<ConnectionPool<P::DbMarker>> {
+        let result = self.builder().set_max_size(size).build().await;
 
         if result.is_ok() {
             let old_count = self.connections_count.fetch_add(size, Ordering::Relaxed);
             let total_connections = old_count + size;
             tracing::info!(
-                "Created a new master pool. Master pool total connections count: {total_connections}"
+                "Created a new {} pool. Total connections count: {total_connections}",
+                P::kind_str()
             );
         }
 
         result
     }
 }
 
-/// Represents a connection pool to the replica database.
 #[derive(Debug, Clone)]
-pub struct ReplicaPoolResource {
-    connections_count: Arc<AtomicU32>,
-    builder: ConnectionPoolBuilder<Core>,
-}
-
-impl Resource for ReplicaPoolResource {
-    fn name() -> String {
-        "common/replica_pool".into()
-    }
-}
-
-impl ReplicaPoolResource {
-    pub fn new(builder: ConnectionPoolBuilder<Core>) -> Self {
-        Self {
-            connections_count: Arc::new(AtomicU32::new(0)),
-            builder,
-        }
-    }
-
-    pub async fn get(&self) -> anyhow::Result<ConnectionPool<Core>> {
-        let result = self.builder.build().await;
+#[non_exhaustive]
+pub struct MasterPool {}
 
-        if result.is_ok() {
-            self.connections_count
-                .fetch_add(self.builder.max_size(), Ordering::Relaxed);
-            let total_connections = self.connections_count.load(Ordering::Relaxed);
-            tracing::info!(
-                "Created a new replica pool. Replica pool total connections count: {total_connections}"
-            );
-        }
-
-        result
-    }
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct ReplicaPool {}
 
-    pub async fn get_singleton(&self) -> anyhow::Result<ConnectionPool<Core>> {
-        self.get_custom(1).await
-    }
+#[derive(Debug, Clone)]
+#[non_exhaustive]
+pub struct ProverPool {}
 
-    pub async fn get_custom(&self, size: u32) -> anyhow::Result<ConnectionPool<Core>> {
-        let result = self.builder.clone().set_max_size(size).build().await;
+pub trait PoolKind: Clone + Sync + Send + 'static {
+    type DbMarker: zksync_db_connection::connection::DbMarker;
 
-        if result.is_ok() {
-            let old_count = self.connections_count.fetch_add(size, Ordering::Relaxed);
-            let total_connections = old_count + size;
-            tracing::info!(
-                "Created a new replica pool. Replica pool total connections count: {total_connections}"
-            );
-        }
-
-        result
-    }
+    fn kind_str() -> &'static str;
```

Comment:
Bikeshedding: would a `const KIND_STR: &'static str` work here? IMO, it looks slightly more idiomatic (ofc, a function is more general, but it's not like implementations are expected to lazily initialize strings, leak them from a `Box<str>` etc.).